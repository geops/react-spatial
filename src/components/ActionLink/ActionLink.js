import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

const propTypes = {
  /**
   * Href attribute.
   */
  href: PropTypes.string,
  /**
   * Children content of the link.
   */
  children: PropTypes.node.isRequired,
  /**
   * Title attribute.
   */
  title: PropTypes.string,
  /**
   * CSS class.
   */
  className: PropTypes.string,
  /**
   * Function trigger on click event.
   */
  onClick: PropTypes.func.isRequired,
};

const defaultProps = {
  className: undefined,
  href: '#',
  title: undefined,
};

/**
 * This component displays a simple a HTML element with an onclick attribute.
 */
class ActionLink extends PureComponent {
  render() {
    const { href, children, title, className, onClick } = this.props;

    return (
      // eslint-disable-next-line jsx-a11y/anchor-is-valid
      <a
        href={href}
        title={title}
        aria-label={title}
        className={className}
        onClick={evt => {
          evt.preventDefault();
          evt.stopPropagation();
          onClick(evt);
        }}
      >
        {children}
      </a>
    );
  }
}

ActionLink.propTypes = propTypes;
ActionLink.defaultProps = defaultProps;

export default ActionLink;
