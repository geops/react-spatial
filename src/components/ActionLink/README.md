#

This demonstrates the use of ActionLink.

```jsx
import React from 'react';
import ActionLink from 'react-spatial/components/ActionLink';

class ActionLinkExample extends React.PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      count: 0,
    };
  }

  render() {
    const { count } = this.state;

    return (
      <div className="tm-action-link-example">
        <ActionLink onClick={() => this.setState({ count: count + 1 })}>
          Click me!
        </ActionLink>
        <div>The action link has been clicked {count} times.</div>
      </div>
    );
  }
}

<ActionLinkExample />;
```
