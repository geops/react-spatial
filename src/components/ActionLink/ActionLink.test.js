import React from 'react';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import ActionLink from './ActionLink';

configure({ adapter: new Adapter() });

describe('ActionLink', () => {
  describe('when no properties are set', () => {
    let spy = null;

    beforeEach(() => {
      window.console.error = jest.fn().mockImplementation(() => {});
      spy = jest.spyOn(window.console, 'error');
    });

    afterEach(() => {
      spy.mockRestore();
      window.console.error.mockRestore();
    });

    test('displays 2 errors in the console', () => {
      shallow(<ActionLink />);
      expect(spy).toHaveBeenCalledTimes(2);
    });

    test('matches snapshot', () => {
      const component = renderer.create(<ActionLink />);
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('when properties are set', () => {
    const label = 'foo';
    const title = ['bar'];
    const className = ['qux'];
    const onClick = () => {};

    test('matches snapshot', () => {
      const component = renderer.create(
        <ActionLink title={title} className={className} onClick={onClick}>
          {label}
        </ActionLink>,
      );
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });

    test('executes onClick property function', () => {
      const fn = jest.fn(onClick);
      const fn2 = jest.fn();
      const evt = {
        preventDefault: fn2,
        stopPropagation: fn2,
      };
      const wrapper = shallow(
        <ActionLink title={title} className={className} onClick={fn}>
          {label}
        </ActionLink>,
      );
      wrapper.simulate('click', evt);
      expect(fn2).toHaveBeenCalledTimes(2);
      expect(fn.mock.calls[0][0]).toBe(evt);
    });
  });
});
