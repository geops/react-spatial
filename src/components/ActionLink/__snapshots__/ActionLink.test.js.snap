// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionLink when no properties are set matches snapshot 1`] = `
<a
  href="#"
  onClick={[Function]}
/>
`;

exports[`ActionLink when properties are set matches snapshot 1`] = `
<a
  aria-label={
    Array [
      "bar",
    ]
  }
  className={
    Array [
      "qux",
    ]
  }
  href="#"
  onClick={[Function]}
  title={
    Array [
      "bar",
    ]
  }
>
  foo
</a>
`;
